engine/api/src/main/scala/pl/touk/nussknacker/engine/api/typed/typing.scala

```diff
@@ -59,42 +59,48 @@ object typing {
     override def objType: TypedClass = TypedClass[Any]
   }
 
-  case class Typed(possibleTypes: Set[TypedClass]) extends TypingResult {
+  // TODO: rename to TypedUnion, ensure that always possibleTypes.size > 1, and cleanup places where was used head/headOption
+  case class Typed(possibleTypes: Set[TypingResult]) extends TypingResult {
+
     override def canBeSubclassOf(typingResult: TypingResult): Boolean = {
       possibleTypes.exists(_.canBeSubclassOf(typingResult))
     }
 
     override def canHaveAnyPropertyOrField: Boolean = {
-      canBeSubclassOf(Typed[util.Map[_, _]]) ||
-        // mainly for avro's GenericRecord purpose
-        hasGetFieldByNameMethod(possibleTypes)
-    }
-
-    private def hasGetFieldByNameMethod(possibleTypes: Set[TypedClass]) = {
-      possibleTypes.exists(tc => tc.klass.getMethods.exists(m => m.getName == "get" && (m.getParameterTypes sameElements Array(classOf[String]))))
+      possibleTypes.exists(_.canHaveAnyPropertyOrField)
     }
 
     override val display : String = possibleTypes.toList match {
       case Nil => "empty"
-      case h::Nil => s"type '${printClass(h)}'"
-      case many => many.map(printClass).mkString("one of (", ", ", ")")
+      case many => many.map(_.display).mkString("one of (", ", ", ")")
     }
 
-    //TODO: should we use simple name here?
-    private def printClass(h: TypedClass) = h.klass.getName
+    override def objType: TypedClass = if (possibleTypes.size == 1) possibleTypes.head.objType else TypedClass[Any]
```

Comment:
Done, but also I realized that `objType = TypedClass[Any]` can works badly because of implementation of `canBeSubclassOf`. For example take a look at `Typer.commonNumberReference` - `Typed[Int].canBeSubclassOf(Typer.commonNumberReference)` will return `false` but opposite condition will return true. Is it ok? It looks like `objType` is covariant but `canBeSubclassOf` is contravariant for union types.

Reply:
I've changed implementation of `canBeSubclassOf` to check "intersection" of subtyping: `firstSet.exists(f => secSet.exists(scalarCanBeSubclassOf(f, _)))`